charon/tests/ui.rs

```diff
@@ -0,0 +1,177 @@
+//! Ui tests for the charon compiler. Each `<file>.rs` file in `ui/` will be passed to the charon
+//! driver. The corresponding pretty-printed llbc output will be stored in `<file>.llbc`, and CI
+//! will ensure these stay up-to-date.
+//!
+//! Files can start with special comments that affect the test behavior. Supported magic comments:
+//! see `HELP_STRING`.
+use anyhow::bail;
+use assert_cmd::prelude::CommandCargoExt;
+use indoc::indoc as unindent;
+use libtest_mimic::{Outcome, Test};
+use std::{
+    error::Error,
+    fs::read_to_string,
+    path::{Path, PathBuf},
+    process::Command,
+};
+use walkdir::{DirEntry, WalkDir};
+
+use charon_lib::cli_options::{CliOpts, CHARON_ARGS};
+
+enum TestKind {
+    PrettyLlbc,
+    KnownPanic,
+    Unspecified,
+}
+
+struct MagicComments {
+    test_kind: TestKind,
+}
+
+static HELP_STRING: &str = unindent!(
+    "Test must start with a magic comment that determines its kind. Options are:
+    - `//@ output=pretty-llbc`: record the pretty-printed llbc;
+    - `//@ known-panic`: a test that is expected to panic.
+    "
+);
+
+fn parse_magic_comments(input_path: &std::path::Path) -> anyhow::Result<MagicComments> {
+    // Parse the magic comments.
+    let mut kind = TestKind::Unspecified;
+    for line in read_to_string(input_path)?.lines() {
+        if !line.starts_with("//@") {
+            break;
+        }
+        let line = line[3..].trim();
+        if line == "known-panic" {
+            kind = TestKind::KnownPanic;
+        } else if line == "output=pretty-llbc" {
+            kind = TestKind::PrettyLlbc;
+        } else {
+        }
```

Comment:
Why not return after we successfully parse the first magic comment?

Reply:
Yeah fair the loop doesn't make sense yet. I wrote it like this because I plan to add other options, e.g. `//@ driver-opts=--no-code-duplication` depending on what comes useful.

Reply:
Makes sense, and it's not that important anyway